pkg/sqlutil/pg/connection.go

```diff
@@ -0,0 +1,122 @@
+package pg
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/XSAM/otelsql"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/stdlib"
+	"github.com/jmoiron/sqlx"
+	"github.com/scylladb/go-reflectx"
+	"go.opentelemetry.io/otel"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+
+	// need to make sure pgx driver is registered before opening connection
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+// NOTE: This is the default level in Postgres anyway, we just make it explicit here.
+const defaultIsolation = sql.LevelReadCommitted
+
+// Driver is a compiler enforced type used that maps to database driver names
+type Driver string
+
+const (
+	// DriverPostgres represents the postgres dialect.
+	DriverPostgres Driver = "pgx"
+	// DriverTxWrappedPostgres is useful for tests.
+	// When the connection is opened, it starts a transaction and all
+	// operations performed on the DB will be within that transaction.
+	DriverTxWrappedPostgres Driver = "txdb"
+)
+
+var (
+	otelOpts = []otelsql.Option{otelsql.WithAttributes(semconv.DBSystemPostgreSQL),
+		otelsql.WithTracerProvider(otel.GetTracerProvider()),
+		otelsql.WithSQLCommenter(true),
+		otelsql.WithSpanOptions(otelsql.SpanOptions{
+			OmitConnResetSession: true,
+			OmitConnPrepare:      true,
+			OmitRows:             true,
+			OmitConnectorConnect: true,
+			OmitConnQuery:        false,
+		})}
+)
+
+type ConnectionConfig struct {
+	DefaultIdleInTxSessionTimeout time.Duration
+	DefaultLockTimeout            time.Duration
+	MaxOpenConns                  int
+	MaxIdleConns                  int
+}
+
+func (config ConnectionConfig) NewDB(ctx context.Context, uri string, driver Driver) (*sqlx.DB, error) {
```

Comment:
I'm trying to understand NewDB as a method on a `ConnectionConfig`, as opposed to it being global for the loop and accepting a connection config. What guided you towards defining it this way?

Reply:
Can you elaborate on the alternative you are referring to?
This is extracted from the logic in core, and repurposed to be used by LOOPPs, but still used by core as well.

Reply:
```
func NewDB(ctx context.Context, config ConnectionConfig, uri string, driver Driver) (*sqlx.DB, error) {
```

Reply:
Typically using a method is less verbose since you don't have to repeat the package and the func doesn't need to name the type.
```go
v := packge.NewThing(package.ThingConfig{})
```
```go
v := package.ThingConfig{}.New()
```
However, in this case the payoff is minor given the short package name `pg` and the difference in config and returned type names.
```go
db, err := pg.ConnectionConfig{...}.NewDB(ctx, uri, driver)
```
```go
db, err := pg.NewDB(ctx, pg.ConnectionConfig{...}, uri, driver)
```
Although sometimes struct at the end is friendlier for inline calls:
```go
db, err := pg.NewDB(ctx, uri, driver, pg.ConnectionConfig{
	Foo: "bar",
	Baz: 42,
})
```
Although, now I am questioning my config type naming - why not call this `DBConfig`? So then we have:
```go
db, err := pg.DBConfig{...}.New(ctx, uri, driver)
```
Or:
```go
db, err := pg.NewDB(ctx, uri, driver, pg.DBConfig{...})
```

Reply:
Updated to `DBConfig.New()`